Fix issue 033506: correct droppedEntriesCount (#33538)
mdn/content: files/en-us/web/api/performanceobserver/performanceobserver/index.md

### Parameters

- `callback`

  - : A `PerformanceObserverCallback` callback that will be invoked when observed performance events are recorded. When the callback is invoked, the following parameters are available:

    - `entries`
      - : The {{domxref("PerformanceObserverEntryList","list of performance observer entries", '', 'true')}}.
    - `observer`
      - : The {{domxref("PerformanceObserver","observer")}} object that is receiving the above entries.
    - `options`

>       - : An object with the following properties:

Reviewer: Maybe optional here?

```suggestion
      - : An optional object with the following properties:
```

Author: If we want to do this we should instead use the `{{optional_inline}}` macro after the parameter name (like this: https://developer.mozilla.org/en-US/docs/Web/API/Request/Request#options).

But, 2 things.

1) I talked about this a bit in the PR description: "optional" is weird when we're talking about callback parameters. When we use "optional" for a normal API method the semantics is clear: developers don't have to supply a value, and if they don't, some default is used. But with a callback, the developers aren't calling it, so how are developers supposed to interpret "optional"? I think it is probably better, rather than "optional", to describe exactly why/under which circumstances the parameter is not passed, so developers know when they can access it.

2) Although this parameter is marked optional in the IDL, I can't understand from the spec how it is optional. AFAICS the relevant bit is https://w3c.github.io/performance-timeline/#queue-the-performanceobserver-task, which has:

> 9. Call po’s [observer callback](https://w3c.github.io/performance-timeline/#dfn-observer-callback) with observerEntryList as the first argument, with po as the second argument and as [callback this value](https://webidl.spec.whatwg.org/#dfn-callback-this-value), and with callbackOptions as the third argument. If this [throws](https://webidl.spec.whatwg.org/#dfn-throw) an exception, [report the exception](https://html.spec.whatwg.org/multipage/webappapis.html#report-the-exception).

I don't see any indication there that `options` may not be passed. It will be empty, for all but the first callback invocation, based on steps 7 and 8. So what am I missing here?

Reviewer: That's fair that it will be empty rather than missing. I guess I'm just trying to avoid the original confusion that was raised in #33506 where the user was confused why it wasn't there as they expected it to contain a value every time (where the `droppedEntriesCount` would be `0` every time after the first—that's not the case). I can understand that confusion. But maybe with all your other changes it's sufficient.

        - `droppedEntriesCount`

>           - : The number of [buffered](/en-US/docs/Web/API/PerformanceObserver/observe#buffered) entries which got dropped from the buffer due to the buffer being full.

Reviewer: Reads weird to me. How can an entry be both buffered and not buffered?

```suggestion
          - : The number of entries which where not included in the [buffered entries](/en-US/docs/Web/API/PerformanceObserver/observe#buffered) due to the buffer being full.
```

Author: Good point. Could we say:

```suggestion
          - : The number of entries which could not be recorded because the {{domxref("Performance")}} object's internal buffer was full.
```

?

Reviewer: Works for me!

>             Note that this is only provided the first time the observer calls the callback, because once the observer starts emitting observations, it can clear the buffer. After the first time, `droppedEntriesCount` will be `undefined`.

Reviewer: The buffer isn't cleared as other PerformanceObservers can be started independently and also need to reuse that buffer.

```suggestion
          Note that this is only provided the first time the observer calls the callback, when the buffered entries are replayed. Once the observer starts observing future observations, it no longer needs to use the buffer. After the first time, `options` will be an empty object (`{}`).
```

I've also unindented so the note applies to `options` but double check my spacing.

Author: Oh, actually, although I merged this suggestion, this is the same issue as elsewhere: on my reading of the spec `options`  is always passed.

>             Note that this is only provided the first time the observer calls the callback, when the buffered entries are replayed. Once the observer starts making future observations, it no longer needs to use the buffer. After the first time, `options` will be an empty object (`{}`).

Author: I think it applies more to `droppedEntriesCount` specifically though. When it says "Note that this is only provided", the referent of "this" is `droppedEntriesCount`, not `options`.

And the second sentence is only about `droppedEntriesCount` as well - we might imagine a future in which other properties get added to `options` (which is presumably the point of making this an object), and they might have nothing to do with buffering, and then tying this sentence to `options` as a whole would look wrong.

>             Note that this is only provided the first time the observer calls the callback, when the buffered entries are replayed. Once the observer starts making future observations, it no longer needs to use the buffer. After the first time, `options` will be an empty object (`{}`).

Reviewer: Yeah agree it's a little ambiguous and kinda applies to both at the moment. It's the last sentence that specifically applies to the `options` section rather than `droppedEntriesCount `.

Anyway, let's go with what's there now.

>             Note that this is only provided the first time the observer calls the callback, when the buffered entries are replayed. Once the observer starts making future observations, it no longer needs to use the buffer. After the first time, `options` will be an empty object (`{}`).

Author: Yeah, I agree, the last sentence still bothers me a bit here. I would happily omit it, but am OK with it like this as well.

>             Note that this is only provided the first time the observer calls the callback, when the buffered entries are replayed. Once the observer starts making future observations, it no longer needs to use the buffer. After the first time, `options` will be an empty object (`{}`).

Author: I agree that the last sentence is a little awkward and would be happy to omit it - it's implicitly true from the other stuff, and has the chance of becoming false if more options are ever added. But I also think it's OK to keep it.

>             Note that this is only provided the first time the observer calls the callback, when the buffered entries are replayed. Once the observer starts observing future observations, it no longer needs to use the buffer. After the first time, `options` will be an empty object (`{}`).

Author: ```suggestion
            Note that this is only provided the first time the observer calls the callback, when the buffered entries are replayed. Once the observer starts making future observations, it no longer needs to use the buffer. After the first time, `options` will be an empty object (`{}`).
```

### Dropped buffer entries

You can use {{domxref("PerformanceObserver")}} with a `buffered` flag to listen to past performance entries.

> There is a buffer size limit, though. The performance observer callback contains an `options` object: the first time the observer calls the callback, the `options` parameter will have a `droppedEntriesCount` property that tells you how many entries were dropped due to the buffer storage being full.

Reviewer: ```suggestion
There is a buffer size limit, though. The first time the observer calls the callback, the performance observer callback may contain an optional `options` object when supported. The `options` parameter will have a `droppedEntriesCount` property that tells you how many entries were dropped due to the buffer storage being full.
```

Author: On my reading of https://w3c.github.io/performance-timeline/#queue-the-performanceobserver-task, steps 6-9, this is not right: `options` is always passed, but after the first time, the `droppedEntriesCount` property is unset.

I'm happy to be corrected but would like to understand why I am wrong.

Reviewer: OK I agree with you. But maybe make it more explicit then?

```suggestion
There is a buffer size limit, though. The performance observer callback contains an `options` object: the first time the observer calls the callback, the `options` parameter will have a `droppedEntriesCount` property that tells you how many entries were dropped due to the buffer storage being full. Subsequent callbacks will have an empty `options` parameter.
```

```js
function perfObserver(list, observer, options) {
  list.getEntries().forEach((entry) => {
    // do something with the entries
  });

>   if (isFirstObservation) {
>     isFirstObservation = false;
>     if (options?.droppedEntriesCount > 0) {
>       console.warn(
>         `${options?.droppedEntriesCount} entries got dropped due to the buffer being full.`,
>       );
>     }
>   }
> }
> 
> const observer = new PerformanceObserver(perfObserver);
> const isFirstObservation = true;

Reviewer: You could just check this each time. Not sure how much you're saving with this `isFirstObservation` variable and reads more complicated to me.

```suggestion
  if (options?.droppedEntriesCount > 0) {
    console.warn(
      `${options?.droppedEntriesCount} entries got dropped due to the buffer being full.`,
    );
  }
}

const observer = new PerformanceObserver(perfObserver);
```

But also happy to keep your code if you prefer.

Author: I agree it is more complicated, but the reason I did that is to communicate in the example _why_ it's sometimes not present. I suppose it's hard to decide whether an example should be what you would actually use, or should demonstrate the API. Anyway, I'm happy either way, just wanted to explain the reasoning.

Reviewer: I'd favour more making the examples reflective of code that would be used as think people will copy and paste these. To me the text rather than the code is the place for the explanation.

But I also don't think it's super critical so happy to leave as is as well.

Author: > I'd favour more making the examples reflective of code that would be used as think people will copy and paste these. To me the text rather than the code is the place for the explanation.

That sounds like the kind of thing I would say :).
